### PR TITLE
s3: Return XMinioInvalidObjectName if the object contains null char

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
GetObject() on a object name with a null character 
will return "drive does not support O_DIRECT "
though the drive supports O_DIRECT. 

The reason is that the code does 
not know for sure why a syscall returns EINVALID 
and it converts it to the unsupported O_DIRECT error, 
which is confusing for users

To avoid this situation, just consider \0 as an invalid 
object character though S3 supports it, but \0 will 
never work with filesystems anyway.

## Motivation and Context
Better error message when \0 is used in a prefix in S3 operations

## How to test this PR?
```
package main

import (
        "context"
        "io"
        "log"
        "os"

        "github.com/minio/minio-go/v7"
        "github.com/minio/minio-go/v7/pkg/credentials"
)

func main() {
        s3Client, err := minio.New("localhost:9000", &minio.Options{
                Creds:  credentials.NewStaticV4("minioadmin", "minioadmin", ""),
                Secure: false,
        }) 
        if err != nil {
                log.Fatalln(err)
        }
        
        reader, err := s3Client.GetObject(context.Background(), "testbucket", "abcd\n\x00", minio.GetObjectOptions{})
        if err != nil {
                log.Fatalln(err)
        }
        defer reader.Close()
                
        if _, err := io.CopyN(os.Stderr, reader, 10); err != nil {
                log.Fatalln(err)
        }
}
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#0ab2428a-0fdf-4f94-84da-0afa58d2d419) |
<a name="0ab2428a-0fdf-4f94-84da-0afa58d2d419"></a>
### 16372-2b004c7/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.4:31414
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp d6e737b52502:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 1 minutes and 1 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 50 seconds
(9/14) Running minio-java tests ... FAILED in 30 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 417,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"173787E99A56FCEA\",\"HostId\":\"f094e82b-7959-45c4-b85f-3a7fa3c5977c\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 49 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 47 seconds
(11/14) Running s3cmd tests ... done in 16 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

File: cmd/object-api-utils.go [hunk 1]
```diff
@@ -170,6 +170,13 @@ func IsValidObjectPrefix(object string) bool {
 	if strings.Contains(object, `//`) {
 		return false
 	}
+	// This is valid for AWS S3 but it will never
+	// work with file systems, we will reject here
+	// to return object name invalid rather than
+	// a cryptic error from the file system.
+	if strings.Contains(object, "\x00") {
```
Comment (klauspost) [inline cmd/object-api-utils.go]:
```suggestion
	if !utf8.Valid(object) {
```
That should be a more generic validation. `import "unicode/utf8"` is probably needed.

Comment (vadmeste) [reply cmd/object-api-utils.go]: @klauspost \x00 is utf8 valid, also we are doing that in line 167

Comment (harshavardhana) [reply cmd/object-api-utils.go:177]: I thought doing this adds latency to the incoming calls or no? - one of the reasons had avoided using utf8 in this codepath.

Comment (klauspost) [reply cmd/object-api-utils.go]: Really? Wow

**Review by klauspost — APPROVED** [review]

File: cmd/object-api-utils.go (hunk 1 above)
Comment (klauspost) [inline cmd/object-api-utils.go]:
```suggestion
	if strings.ContainsRune(object, 0) {
```

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**